File: benchmarks3d/joint_fixed3.rs
use rapier_testbed3d::Testbed;
use rapier3d::prelude::*;

pub fn init_world(testbed: &mut Testbed) {
    /*
     * World
     */
    let mut bodies = RigidBodySet::new();
    let mut colliders = ColliderSet::new();
    let mut impulse_joints = ImpulseJointSet::new();
    let multibody_joints = MultibodyJointSet::new();

    let rad = 0.4;
    let num = 5;
    let shift = 1.0;

    let mut body_handles = Vec::new();

    for m in 0..10 {
        let z = m as f32 * shift * (num as f32 + 2.0);

        for l in 0..10 {
            let y = l as f32 * shift * 3.0;

            for j in 0..5 {
                let x = j as f32 * shift * (num as f32) * 2.0;

                for k in 0..num {
                    for i in 0..num {
                        let fk = k as f32;
                        let fi = i as f32;

                        // NOTE: the num - 2 test is to avoid two consecutive
                        // fixed bodies. Because physx will crash if we add
                        // a joint between these.

                        let status = if i == 0 && (k % 4 == 0 && k != num - 2 || k == num - 1) {
                            RigidBodyType::Fixed
                        } else {
                            RigidBodyType::Dynamic
                        };

                        let rigid_body = RigidBodyBuilder::new(status).translation(vector![
                            x + fk * shift,
                            y,
                            z + fi * shift
                        ]);
                        let child_handle = bodies.insert(rigid_body);
                        let collider = ColliderBuilder::ball(rad);
                        colliders.insert_with_parent(collider, child_handle, &mut bodies);

                        // Vertical joint.
                        if i > 0 {
                            let parent_handle = *body_handles.last().unwrap();
                            let joint =
                                FixedJointBuilder::new().local_anchor2(point![0.0, 0.0, -shift]);
                            impulse_joints.insert(parent_handle, child_handle, joint, true);
                        }

                        // Horizontal joint.
                        if k > 0 {
                            let parent_index = body_handles.len() - num;
                            let parent_handle = body_handles[parent_index];
                            let joint =
                                FixedJointBuilder::new().local_anchor2(point![-shift, 0.0, 0.0]);
                            impulse_joints.insert(parent_handle, child_handle, joint, true);
                        }

                        body_handles.push(child_handle);
                    }
                }
            }
        }
    }

    /*
     * Set up the testbed.
     */
    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints);
    testbed.look_at(point![-38.0, 14.0, 108.0], point![46.0, 12.0, 23.0]);
}

File: python/rapier-testbed/rapier_testbed/examples3/stress_tests/joint_fixed3.py
"""Port of examples3d/stress_tests/joint_fixed3.rs."""
from __future__ import annotations

import rapier3d as rp
from ..._registry import register

CATEGORY = "Stress Tests"
NAME = "ImpulseJoint fixed"


def init_world(testbed) -> None:
    bodies = rp.RigidBodySet()
    colliders = rp.ColliderSet()
    impulse_joints = rp.ImpulseJointSet()
    multibody_joints = rp.MultibodyJointSet()

    rad = 0.4
    num = 5
    shift = 1.0
    body_handles = []

    for m in range(10):
        z0 = m * shift * (num + 2.0)
        for l in range(10):
            y0 = l * shift * 3.0
            for jj in range(5):
                x0 = jj * shift * num * 2.0
                for k in range(num):
                    for i in range(num):
                        fk = float(k)
                        fi = float(i)
                        if i == 0 and ((k % 4 == 0 and k != num - 2) or k == num - 1):
                            bb = rp.RigidBody.fixed()
                        else:
                            bb = rp.RigidBody.dynamic()
                        bb = bb.translation((x0 + fk * shift, y0, z0 + fi * shift))
                        child = bodies.insert(bb)
                        colliders.insert_with_parent(rp.Collider.ball(rad), child, bodies)

                        if i > 0:
                            parent = body_handles[-1]
                            j = rp.FixedJoint.builder().local_anchor2((0.0, 0.0, -shift)).build()
                            impulse_joints.insert(parent, child, j, wake_up=True)
                        if k > 0:
                            parent = body_handles[len(body_handles) - num]
                            j = rp.FixedJoint.builder().local_anchor2((-shift, 0.0, 0.0)).build()
                            impulse_joints.insert(parent, child, j, wake_up=True)
                        body_handles.append(child)

    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints)
    testbed.look_at((-38.0, 14.0, 108.0), (46.0, 12.0, 23.0))


register(CATEGORY, NAME, init_world, dim=3)


if __name__ == "__main__":
    from ..._picker import run

    run(initial=f"{CATEGORY} / {NAME}")
